### PR TITLE
[WIP] Ubershaders

### PR DESCRIPTION
**Pull request by phire**
## Ubershaders: A long term project.

The ultimate goal of ubershaders is to completely eliminate all stuttering caused by shader compilation. There are a few possible methods in which ubershaders could be used to elimiate stuttering. 

Ideally we would only need one ubershader, but it's likely we will end up with at least 8 shaders to take advantage of features that can only be changed at compile time. 
1. It is entirely possible that ubershaders (once completed) will be fast enough on modern superscaler GPUs to play games without needing to generating unique shaders for each TEV state. Ubershaders play right into the strengths of SMT, as every thread executing the same TEV state follows the exact same pattern of branches. 
2. Ubershaders can be combined with async shader compilation of our current xShaderGen shaders. When a new TEV state is encountered we generate the shader and queue it for compilation with the drivers. In parallel we bind the ubershader and render the current object. Next time we encounter the same TEV state, we check to see if the driver has finished compiling the xShaderGen shader and use that.
3. We can even mix and match the two methods, keep track of which TEV states render the most fragments and only queue them for compilation once they have crossed a threshold. Or put the thread rendering the shaders at a low priority and whenever it finishes compiling one shader feed it the next most used shader.
### How long does Shader compilation take.

Too long. Measurements by JMC47 put the average shader over 10ms, with many shaders taking 45ms or even over 100ms.

As we only have a budget of 16-32ms to render the entire frame, we can't really afford the hit of compiling shaders. Often a new object popping up on the screen, or a new effect, or moving into a new room will cause lots of stuttering as many new TEV states are encountered and many shaders need to be compiled.
### This PR.

The goal of this PR is to implement basic ubershaders so the performance of the concept can be tested.
### Current status

Currently it implements just enough to get helloTriangle working, though it could do upto 16 stages of tev.

image

Generally nothing else uses such a limited set of TEV, but things which are pure vertex colors, like Mario in Mario 64 work (though the level appears to be texture based)

naap01-1

Only pixel shaders are implemented. It uses VertexShaderGen for vertex shaders. It doesn't try anything fancy yet and compiles exactly as many shader programs as regular dolphin. And it's only implemented in OpenGL so far (need to implement bitfieldExtract on directX).

Next on my list of things to do:
- [x] Basic textures (not indirect)
- [x] Konst
- [x] Alpha Combiner
- [x] Color channel swapping (swizzling)
- [x] Alpha Test
- [x] DirectX support
- [x] Fog
- [ ] Vertex UberShader
- [x] Don't recompile the same ubershaders for every UID.
- [ ] Dummy indirect texture loads (so we can get an approximate benchmark)

I think that will be enough for this PR, should give use a reasonable idea of expected performance.

[Review on Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3163)

### CODE REVIEW

File: Source/Core/VideoCommon/BPMemory.h [hunk 1]
```diff
@@ -382,28 +378,23 @@ struct TevStageCombiner
...
-		};
 		u32 hex;
-		int getTexMap(int i){return i?texmap1:texmap0;}
-		int getTexCoord(int i){return i?texcoord1:texcoord0;}
-		int getEnable(int i){return i?enable1:enable0;}
-		int getColorChan(int i){return i?colorchan1:colorchan0;}
+		u32 getTexMap(int i) { return i ? u32(texmap1) : u32(texmap0); }
+		u32 getTexCoord(int i) { return i ? u32(texcoord1) : u32(texcoord0); }
+		u32 getEnable(int i) { return i ? u32(enable1) : u32(enable0); }
+		u32 getColorChan(int i) { return i ? u32(colorchan1) : u32(colorchan0); }
```
Comment (lioncash) [inline Source/Core/VideoCommon/BPMemory.h]: You can just call .Value() instead of casting.

File: Source/Core/VideoCommon/UberShaderPixel.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2015 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include "VideoCommon/PixelShaderGen.h"
+
+namespace UberShader {
```
Comment (lioncash) [inline Source/Core/VideoCommon/UberShaderPixel.h]: brace on new line

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+#include "VideoCommon/UberShaderPixel.h"
+
+namespace UberShader
+{
+
+static char text[32768];
+
+template<typename T>
+std::string bitfieldExtract(std::string source, T type)
```
Comment (lioncash) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Capital B, also const reference.

File: Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp [hunk 4]
```diff
@@ -157,6 +162,9 @@ void ProgramShaderCache::UploadConstants()
 		memcpy(buffer.first + ROUND_UP(sizeof(PixelShaderConstants), s_ubo_align) + ROUND_UP(sizeof(VertexShaderConstants), s_ubo_align),
 			&GeometryShaderManager::constants, sizeof(GeometryShaderConstants));
 
+		memcpy(buffer.first + ROUND_UP(sizeof(PixelShaderConstants), s_ubo_align) + ROUND_UP(sizeof(VertexShaderConstants), s_ubo_align) + ROUND_UP(sizeof(GeometryShaderConstants), s_ubo_align),
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp]: I don't think it's wise to include this buffer within the other uniforms. So it must be reuploaded on all uniform changes. It's likely better to create a new uniform buffer object for every UID, and to just bind it on every change.

Comment (phire) [inline Source/Core/VideoBackends/OGL/ProgramShaderCache.cpp]: Yeah, that was my "just bashing things through so I could try this before sleeping" effort.

**Comment by phire** [comment]
### Progress Update

I've added support for Konst and Basic Texturing. 

#### Screenshots:
Wind Waker's title screen looks great as always:
image

I think if I added Alpha support, Mario 64 would be more or less playable:
image

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+#include "VideoCommon/UberShaderPixel.h"
+
+namespace UberShader
+{
+
+static char text[32768];
+
+template<typename T>
+std::string BitfieldExtract(const std::string &source, T type)
```
Comment (lioncash) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: ampersand against the type

**Comment by phire** [comment]
### Progress Update:

Implemented Alpha combiner and dual source blending. Something is still missing, because the trees in Super Mario 64 still have issues, but the shadow is correct now.

naap01-3

However, the Intro to Wind Waker (and most other things) has improved a lot.

gzle01-4

Tomorrow I think I'll implement color channel swapping, as I suspect this is the cause of a few issues.

**Comment by phire** [comment]
### Progress update

I've added support for ubershaders both OpenGL 3.3 and earlier (mostly to get fifoci working) and DirectX. 

Also, I added Alpha test, which has fixed most of the remaining alpha issues. N64 games now seem to render more or less perfectly (I haven't looked for that hard for missing effects)

Super Mario 64:
naap01-6

**Comment by JMC47** [comment]
Mario 64 is missing fog, but I think that's known.

**Comment by phire** [comment]
### Progress Update

FifoCI was taking a lot time to run with ubershaders (runtime had jumped from 6 to 32min), so I made changes so it wouldn't generate and compile duplicate ubershdaers. This drops fifoci runtime down to a much more acceptable 24min (yeah, it still needs improvement)

In other (completely related) news, this gets rid of a lot of shader compiler based stuttering. It still has to generate vertex shaders, and upto 16 pixel uber shaders as it encountered them, but it's a lot smoother.

Average frame time and fps before this change:
worse

And after. Notice how the time per frame axis is now in milliseconds instead of instead of seconds:
better

File: Source/Core/VideoCommon/BPMemory.h [hunk 6]
```diff
@@ -302,41 +302,37 @@ struct TevStageCombiner
...
-
-			u32 bias : 2; //GXTevBias
-			u32 op : 1;
-			u32 clamp : 1;
-
-			u32 shift : 2;
-			u32 dest : 2;  //1,2,3
-		};
+		BitField<0,  2, u32> rswap;
+		BitField<2,  2, u32> tswap;
```
Comment (degasus) [inline Source/Core/VideoCommon/BPMemory.h]: This needs some (int) casts in https://github.com/phire/dolphin/blob/ubershaders/Source/Core/VideoCommon/BPStructs.cpp#L1283

File: Source/Core/VideoCommon/PixelShaderManager.cpp [hunk 7]
```diff
@@ -280,6 +321,39 @@ void PixelShaderManager::SetFogRangeAdjustChanged()
...
+	}
+	else if (bp == 0xf3)
+	{
+		more_constants.alphaTest = newValue;
+		dirty = true;
+	}
+	else if (bp >= 0xf8)
+	{
+		u32 ksel = bp - 0xf8;
+		more_constants.tevksel[ksel][0];
```
Comment (degasus) [inline Source/Core/VideoCommon/PixelShaderManager.cpp]: = newValue?

Comment (phire) [inline Source/Core/VideoCommon/PixelShaderManager.cpp]: Yeah, that could help.

**Comment by phire** [comment]
@lioncash Can I get your opinion on those changes to BitField (the 6 fixup commits)

**Comment by lioncash** [comment]
@phire What was the reasoning for changing the constexpr BitField functions to static const vars?

**Comment by phire** [comment]
@lioncash Is it bad if I say "So I didn't have to type ()"

My reasoning was this was this would be slightly cleaner.

**Comment by phire** [comment]
### Progress Update

#### Performance

Spent a lot of yesterday optimizing the ubershaders. The dynamically indexed ColorInput/AlphaInput selection arrays which we updated and read back each loop though the tev stages were being stored in main memory. So the shader was inherently L2 cache and memory bound. On my GTX 960, it worked well for Wind Waker at 1x IR, struggled a little at 1.5xIR and had large performance drops at 2xIR, where it was writing over 2GB of to main memory each frame and reading back 550MB.

By re-arranging the shader to use select statements instead of dynamically indexed arrays (An optimization I really think shader compilers should be able to do themselves) all the values stay in registers (though, there is a massive increase in uniform control flow, which is way faster) and the shaders go a lot faster.

OpenGL on my GTX 960 can now manage that same scene in wind waker at 3xIR with a slight drop to 27fps, and DirectX can manage an impressive 5xIR at the full 30fps. I'm not really sure why Nvidia's drivers are generating code for DirectX which is about 3 times faster than it generates for OpenGL. Would help if I could see the actual shader assembly, but none of nvidia's tools make that available. I'll look into that more later.

#### Accuracy

@degasus also noticed a major bug with konst selection. Fixing that makes Wind Waker intro render almost correctly. The clouds at the back pop-in and don't look correct, which I assume is due to the lack of fog emulation.
gzle01-7


In game, the cell shading of the characters, resulting in red gradients (see Link's hair) instead of the expected lighting. This is due to the lack of emulation of Tev's compare mode.
gzle01-5

**Comment by lioncash** [comment]
@phire I don't particularly mind either way (though I have a preference for functions, outside of type trait-likes), but they should still be made constexpr for use in other constant expressions.

**Comment by phire** [comment]
The size and offset of a bitfield is a type trait.

**Comment by lioncash** [comment]
A static variable primitive and a type-trait struct are very different things. Note that STL types don't expose a variable behind size(), neither does BitSet (for its members) or FifoQueue.

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 8]
```diff
@@ -0,0 +1,687 @@
...
+		"			int3 color;\n"
+		"			if(color_bias != 3u) { // Normal mode\n"
+		"				color.r = tevLerp(color_A.r, color_B.r, color_C.r, color_D.r, color_bias, color_op, color_shift);\n"
+		"				color.g = tevLerp(color_A.g, color_B.g, color_C.g, color_D.g, color_bias, color_op, color_shift);\n"
+		"				color.b = tevLerp(color_A.b, color_B.b, color_C.b, color_D.b, color_bias, color_op, color_shift);\n"
+		"			} else { // Compare mode\n"
+		"				// op 6 and 7 do a select per color channel\n"
+		"				if (color_compare_op == 6u) {\n"
+		"					// TEVCMP_RGB8_GT\n"
+		"					color.r = (color_A.r > color_B.r) ? color_D.r : 0;\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: I guess s/color_D/color_C/ ?

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 9]
```diff
@@ -0,0 +1,687 @@
...
+		"}\n\n");
+
+	// =======================
+	//   TEV's Color Compare
+	// =======================
+
+	out.Write(
+		"// Implements operations 0-5 of tev's compare mode,\n"
+		"// which are common to both color and alpha channels\n"
+		"bool tevCompare(uint op, int3 color_A, int3 color_B) {\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Do you think it's worth to return a bool4 instead? So also case 6u and 7u could be handled.

File: Source/Core/VideoCommon/BPStructs.cpp [hunk 10]
```diff
@@ -302,6 +305,7 @@ static void BPWritten(const BPCmd& bp)
 	case BPMEM_FOGPARAM3:
 		if (bp.changes)
 			PixelShaderManager::SetFogParamChanged();
+		PixelShaderManager::UpdateBP(bp.address, bp.newvalue);
```
Comment (degasus) [inline Source/Core/VideoCommon/BPStructs.cpp]: Please just create a new PixelShaderManager::SetFogParam() function.

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 11]
```diff
@@ -0,0 +1,687 @@
...
+	//        Should be fixed point, and should not make guesses about Range-Based adjustments.
+	out.Write(
+		"	// Fog\n"
+		"	uint fog_function = %s;\n", BitfieldExtract("bpmem_fogParam3", FogParam3().fsel).c_str());
+	out.Write(
+		"	if (fog_function != 0u) {\n"
+		"		// TODO: This all needs to be converted from float to fixed point\n"
+		"\n"
+		"		// TODO: zCoord is hardcoded to fast depth with no zfreeze\n"
+		"		int zCoord = " I_ZBIAS"[1].x + int((clipPos.z / clipPos.w) * float(" I_ZBIAS"[1].y));\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: iirc we default to fastdepth, so int zCoord = int(gl_FragCoord.z * 16777215.0);

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+		"\n");
+	// Use dual-source color blending to perform dst alpha in a single pass
+	if (dstAlphaMode == DSTALPHA_DUAL_SOURCE_BLEND)
+	{
+		// Colors will be blended against the alpha from ocol1 and
+		// the alpha from ocol0 will be written to the framebuffer.
+		out.Write(
+			"	// dual source blending\n"
+			"	ocol1 = float4(TevResult) / 255.0;\n"
+			"	ocol0.a = float(" I_ALPHA".a) / 255.0;\n"
```
Comment (degasus) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: hey, inlining this is quite simple: You are always allowed to write to col1.

**Comment by JMC47** [comment]
Videos in most games are still rendering with a red tint, pointing to a bug.

File: Source/Core/VideoCommon/BPMemory.h [hunk 13]
```diff
@@ -692,20 +683,17 @@ union FogParam0
...
-		u32 c_exp  : 8;
-		u32 c_sign : 1;
-		u32 proj   : 1; // 0 - perspective, 1 - orthographic
-		u32 fsel   : 3; // 0 - off, 2 - linear, 4 - exp, 5 - exp2, 6 - backward exp, 7 - backward exp2
-	};
+	BitField<0, 11, u32> c_mant;
+	BitField<11, 8, u32> c_exp;
+	BitField<19, 1, u32> c_sign;
+	BitField<20, 1, u32> proj; // 0 - perspective, 1 - orthographic
+	BitField<21, 3, u32> fsel; // 0 - off, 2 - linear, 4 - exp, 5 - exp2, 6 - backward exp, 7 - backward exp2
```
Comment (degasus) [inline Source/Core/VideoCommon/BPMemory.h]: This needs an integer cast in PixelShaderGen.cpp:1154

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 14]
```diff
@@ -0,0 +1,707 @@
...
+
+namespace UberShader
+{
+
+static char text[32768];
+
+template<typename T>
+std::string BitfieldExtract(const std::string& source, T type)
+{
+	return StringFromFormat("bitfieldExtract(%s, %u, %u)", source.c_str(), u32(type.offset), u32(type.size));
```
Comment (lioncash) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: I don't think it's necessary to cast offset and size anymore here.

File: Source/Core/VideoCommon/ShaderGenCommon.h [hunk 15]
```diff
@@ -242,33 +242,37 @@ static void DefineOutputMember(T& object, API_TYPE api_type, const char* qualifi
...
 	DefineOutputMember(object, api_type, qualifier, "float4", "pos", -1, "POSITION");
 	DefineOutputMember(object, api_type, qualifier, "float4", "colors_", 0, "COLOR", 0);
 	DefineOutputMember(object, api_type, qualifier, "float4", "colors_", 1, "COLOR", 1);
 
-	for (unsigned int i = 0; i < xfmem.numTexGen.numTexGens; ++i)
-		DefineOutputMember(object, api_type, qualifier, "float3", "tex", i, "TEXCOORD", i);
+	//for (unsigned int i = 0; i < 8; ++i)
+	if (num_texgens != 0)
+	{
+		std::string tex = StringFromFormat("tex[%d]", num_texgens);
```
Comment (lioncash) [inline Source/Core/VideoCommon/ShaderGenCommon.h]: num_texgens' formatter should be %u

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 16]
```diff
@@ -0,0 +1,707 @@
...
+		"	int AlphaBump = 0;\n"
+		"	int4 icolors_0 = iround(colors_0 * 255.0);\n"
+		"	int4 icolors_1 = iround(colors_1 * 255.0);\n"
+		"	int4 TevResult;\n"
+		"	State s;\n"
+		"	s.TexColor = int4(0, 0, 0, 0);\n"
+		"	s.RasColor = int4(0, 0, 0, 0);\n"
+		"	s.KonstColor = int4(0, 0, 0, 0);\n"
+		"\n");
+	for (int i = 0; i < 4; i++) out.Write(
```
Comment (lioncash) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: body on next line

**Comment by Sonicadvance1** [comment]
171040 <@HdkR> I feel like some vendor is going to break with the dual source blending commit
171043 <@HdkR> Some mobile vendor
171143 <@HdkR> Since it declares two outputs in a row. The spec says if there is one then it defaults to zero, but an array outputs would default to 0-n output locations.
171220 <@HdkR> But it isn't clear on what it should do when you declare multiple outputs that aren't in an array, be it that the outputs increment output draw buffers for each declared or what
171247 <@HdkR> If each default to location zero then it'll result in an error
171319 <@HdkR> So I feel like some mobile vendor is going to fall on its face there ;)
171417 <@HdkR> On desktop GL with dual source blending it doesn't result in an issue because we always bind an indexed location properly

**Comment by phire** [comment]
@Sonicadvance1 Should we be changing it to an array?

**Comment by Sonicadvance1** [comment]
No. We only have one draw buffer even with dual source blending the ocol1 is still bound to draw buffer 0.

**Comment by phire** [comment]
Ok, When I implement the non-dual source blending fallback this issue will be fixed.

File: Source/Core/VideoCommon/UberShaderPixel.cpp [hunk 17]
```diff
@@ -0,0 +1,792 @@
...
+			else if (ApiType == API_D3D) out.Write(
+				"	case %du: return int4(Tex[%d].Sample(samp[%d], float3(uv, 0.0)) * 255.0);\n", i, i, i);
+		}
+		out.Write(
+			"	}\n"
+			"}\n\n");
+	}
+
+	// ======================
+	//   Arbatary Swizzling
```
Comment (Armada651) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Just nitpicking, should be Arbitrary Swizzling.

Comment (phire) [inline Source/Core/VideoCommon/UberShaderPixel.cpp]: Nitpick away

File: Source/Core/VideoCommon/UberShaderPixel.h [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+namespace UberShader
+{
+
+#pragma pack(1)
+struct pixel_ubershader_uid_data
+{
+	// Nice and simple
+
+	// This is the current state, not to be confused with the final state.
+	// Currently: 16 diffrent ubershaders
```
Comment (Armada651) [inline Source/Core/VideoCommon/UberShaderPixel.h]: diffrent -> different

**Comment by phire** [comment]
### Progress Update.

I haven't done a progress update since last week.

Since then I've implemented a number of smaller features and bug fixes. 

 * Arbitrary Texture/Ras Color Swizzling fixes the character cell shading in Wind Waker.
 * An 'off by two' bug in Konst colors fixes the black areas around the edge of the water in Wind Waker, along with countless other bugs.
 * The missing clouds in the distance turned out to be a bug with which set of texture coordinates we were using,
 
With these features/fixes, Wind Waker now renders close to perfectly, as long as you stay away from heat/fire effects (which are indirect textures)
gzle01-18

 * Zfreeze was re-implemented, so are lot of the effect in RS2 are working, including self-shadowing,

gswe64-2

* Depth textures fixes many games (like Mario Kart Wii) which copy part of the framebuffer away to do bloom effects then restore it with depth textures. Without depth textures you are left with a broken square.

 * Fog has been implemented. Not that visible in most games, but noticeable. 

image

#### Known bugs

Check the [fifoci runs](https://fifoci.dolphin-emu.org/version/2380bf1946c377abd135f25037092fada4ccb7d6/) 

There seems to be a small issue with texture coordinates, which is causing small off-by-one bit issues. 

Right now I'm ignoring the bugs and working on vertex shaders instead.

#### Vertex Shaders

I'm doing these over in PR #3185 so I can get separate fifoci runs without the outstanding pixel ubershader bugs getting in the way. Current status is:

No Hardware lighting, no texture coordinates, but some games work quire well with these limiations.

Mario 64 is quite playable, if you remember the location of the now invisible trees:
naap01-8

**Comment by DonelBueno** [comment]
Any news?

**Comment by nbohr1more** [comment]
Rebase please? :)

**Comment by phire** [comment]
Sorry that took a while. Rebased.

**Comment by shuffle2** [comment]
why is fifoci only detecting changes on ogl?

**Comment by phire** [comment]
This doesn't touch video software and the DirectX fifoci hasn't run yet.

**Comment by DonelBueno** [comment]
Will this be pushed into 5.0?

**Comment by Sonicadvance1** [comment]
No

**Comment by phire** [comment]
> No

Awww

**Comment by phire** [comment]
@dolphin-emu-bot rebuild

**Comment by Fallcrest** [comment]
5.0 is out! Insta-merge this NOW! Finally, all rules are out the window!

(More seriously, why is there no Linux build available for 5.0?)

I will be closely following this PR now that work can start back up on it. Exciting stuff.

**Comment by Hamodebu50** [comment]
When will this be implemented in the master branch?

**Comment by Hamodebu50** [comment]
can anyone upload a build of this PR? I Want to try it

**Comment by Pringo** [comment]
@Hamodebu50 

https://dl.dolphin-emu.org/prs/pr-3163-dolphin-latest-dbg-x64.7z
https://dl.dolphin-emu.org/prs/pr-3163-dolphin-latest-x64.7z
https://dl.dolphin-emu.org/prs/pr-3163-dolphin-latest.apk
https://dl.dolphin-emu.org/prs/pr-3163-dolphin-latest.dmg

**Comment by JosJuice** [comment]
@Hamodebu50 Don't download the dbg one, though.

**Comment by ppmeis** [comment]
Thanks for the builds. I just tested with Mario Kart Wii and the improvement is amazing! Great job! Just curious, but this is not implemented under Directx12 right? I tested it and there's no improve at all...Directx11 works perfect.

Thanks @phire for your work.

**Comment by Hamodebu50** [comment]
Shadow the Hedgeho: Menus
Before: 30FPS 
After: 9-11FPS

**Comment by Fallcrest** [comment]
Testing!
(5.0 was at 60 FPS the whole time, OpenGL backend)
Kirby Air Ride opening: 
20 FPS

Menu:
23 FPS

In game:
12-39 FPS (averaging around 27)

Major slowdowns, of course, but graphically, everything looked to be in order. The graphics are pretty simple so I wasn't expecting anything wrong.

Wario World menu:
27 FPS

In Game:
5-31 FPS (When actually moving, it gets extremely slow, but just standing still is around 25)

More complicated graphics including lens flares, shadows and translucent things, but everything seemed fine, except of course for the frame rate, which was unplayable.

Super Mario Sunshine opening:
30 FPS (dipped down to 25 for a little bit near the end)

Title Screen:
15 FPS

In game:
14 FPS or less.

Again, everything seemed to be in order.

Pitfall Lost Expedition opening:
30-60 (Actually watchable, which was surprising)

Menu:
56-60 FPS

In game:
6-7 FPS

Everything again looked fine.

Well done, Phire. Despite major lag and failing a great number of bot tests, everything I tried looked fine. Hope this helps in some way.

**Comment by Hamodebu50** [comment]
just needs some optimization in speed

**Comment by JMC47** [comment]
Performance testing this right now is pretty pointless; this isn't how it will work in master...

**Comment by Hamodebu50** [comment]
MERGE. IT. NOW.

**Comment by Helios747** [comment]
closed in favor of https://github.com/dolphin-emu/dolphin/pull/5702

**Pull request closed without merging**